src/safir/dependencies/db_session.py

```diff
@@ -109,10 +110,29 @@ async def initialize(
             failed transactions due to a non-default isolation level.)
         """
         self._manage_transactions = manage_transactions
-        self._engine = create_database_engine(
-            url, password, isolation_level=isolation_level
-        )
-        self._session = await create_async_session(self._engine)
+        if self._override_engine:
+            self._session = await create_async_session(self._override_engine)
+        else:
+            self._engine = create_database_engine(
+                url, password, isolation_level=isolation_level
+            )
+            self._session = await create_async_session(self._engine)
+
+    def override_engine(self, engine: AsyncEngine) -> None:
+        """Force the dependency to use the provided engine.
+
+        Intended for testing, this allows the test suite to configure a single
+        database engine and share it across all of the tests, benefiting from
+        connection pooling for a test speed-up.  In the Gafaelfawr test suite,
+        sharing an engine for all tests improved the time it took to run the
+        full test suite by XX%.
```

Comment:
Do you have the speed-up statistic available? Was it something like 10%?

Reply:
It was about 5%, and there's a lot of noise. It was just borderline enough to keep, but I removed the claim of a specific speedup and replaced it with a note that it's probably not worth it unless you have a really long test suite.